Implement BPF-based masquerading for IPv6 (rebased)

qmonnet opened the pull request:
Summary of commits:
1. Preparatory: Refactor `ct_is_reply4` as a macro to facilitate implem. of IPv6 counterpart.
2. Preparatory (verifier): Change `ipv6_addrcmp()` into `ipv6_addr_equals()`.
3. Preparatory (verifier): Make `DEFINE_U*()` macros use `.data` section for declarations.
4. Preparatory: Rename `ENABLE_MASQUERADE` as `ENABLE_MASQUERADE_IPV4`.
5. Implement the datapath changes by following the IPv4 implementation.
6. Add ICMP support to the new `ct_extract_ports6()`, refactor to get closer to IPv4's version.
7. Add `ENABLE_MASQUERADE_IPV6` to options lists for complexity tests.
8. Preparatory: Refactor `InitBPFMasqueradeAddrs` to facilitate implem. of IPv6 counterpart.
9. Implement agent changes for the flags and macros.
10. Forbid IPv6 BPF masquerading with host firewall, until we look deeper into the datapath aspects.
11. Update documentation.
12. 13\. 14\. Extend/adjust the tests to cover IPv6 BPF masquerading.

See commits for details.

Fixes: https://github.com/cilium/cilium/issues/14350.

```release-note
Implement BPF-based masquerading for IPv6
```

This is a rebase of https://github.com/cilium/cilium/pull/18694 from Paul.

pchaigno committed:
daemon, config: Enable IPv6 BPF masquerading

This commit allows enabling BPF masquerading for IPv6 and defines all
the proper macros for the datapath implementation.

[ Quentin: Rebased on main branch, addressed minor conflicts on
  daemon.go, daemon_main.go, loader.go, and address.go (including from
  commit d10e8ebed386 ("bpf: dynamic-width static data inlining")),
  adjusted to the changes introduced in previous commits. ]

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno committed:
bpf: Implement IPv6 BPF masquerading

This commit implements BPF masquerading for IPv6 by following the
implementation for IPv4. The main function changed here is
snat_v6_needed with almost the exact same steps as IPv4.

Support for the ipmasq agent in IPv6 is not implemented here.

[ Quentin: Addressed feedback from latest reviews, rebased on main
  branch. Used ENABLE_MASQUERADE_IPV6 instead of sharing
  ENABLE_MASQUERADE with IPv4. Addressed conflicts in nodeport.h and
  nat.h due to commits:
  - 0ea434115253 ("bpf: Make union v6addr const where possible")
  - c62b2094dbec ("bpf: nat: move snat_v{4,6}_needed into lib/nat.h")
  - b3747c8a1bf0 ("bpf: Change ipcache lookup functions to take ClusterID")
  - 9f6fc7e16df9 ("bpf: nodeport: fix tracing for handle_nat_fwd()")
  - 54a86312782f ("bpf: nodeport: handle revDNAT for local backends at to-netdev/to-overlay")
  - eee902f44979 ("bpf: conntrack: introduce an optimized CT lookup for LB")
  - 05d084baeef1 ("bpf: nodeport: move DSR handling from bpf_lxc into from/to-netdev")
  - 306c2b4c83a6 ("bpf: nodeport: Extract DSR specific option from the Geneve header")
  - d749a590069f ("bpf: Pass ext_err to ct_create4 and ct_create6")
  - 8b6aa6eda919 ("bpf: nodeport: fix up trace point in to-overlay NAT paths")
  - 2ec713e8e3c9 ("bpf: rename remote_endpoint_info.sec_label to sec_identity")
  - 5ba20bd7e7e5 ("bpf: Use ct_extract_ports4 for ct_lookup4")
  - "bpf: Change ipv6_addrcmp() into ipv6_addr_equals()" from same PR
]

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno committed:
test: Test BPF masquerading for IPv6

Since Cilium supports BPF masquerading for IPv6, we can now test it.

The upgrade/downgrade test needs a small exception because IPv6 BPF
masquerading is still not supported on v1.13, from which we upgrade.
The exception logic comes with an assertion, to ensure we don't forget
to remove the exception once the upgrade is performed from v1.14.

[ Quentin:
    - Bump version number requirement in assert from 1.12 to 1.14,
      switch fixed assert to versioncheck.
    - Drop test updates for test/k8s/datapath_configuration.go, no
      longer relevant after commits b5e0e124226e ("test: Remove
      K8sDatapathConfig some direct routing tests") and 84116274b4b4
      ("test: Remove K8sDatapathConfig some tunneling tests").
      We don't even have to enable IPv6 masquerading in the datapath
      conformance tests, because of commit 79aa8b1efa7c ("gh/workflows:
      re-enable masqueranding for EGW"). ]

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno committed:
bpf: Refactor ct_is_reply as a macro

This will allow us to easily define the IPv6 equivalent in a subsequent
commit. This commit should include no functional changes.

[ Quentin: Addressed minor conflict with commit 5ba20bd7e7e5 ("bpf: Use
  ct_extract_ports4 for ct_lookup4"). ]

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno committed:
node: Refactor InitBPFMasqueradeAddrs

This commit refactors InitBPFMasqueradeAddrs to easily extend it for
IPv6 in a subsequent commit. This commit shouldn't have any functional
changes.

[ Quentin: Rebased on main branch, addressed minor conflicts. Also:
    - Added firstGlobalAddr() stub to address_other.go, given that it is
      now necessary to build integration tests for Darwin.
    - Moved initMasqueradeAddrs (and V4 version) to address_linux.go
      instead of assigning it in a variable in InitBPFMasqueradeAddrs,
      because of the use of netlink.FAMILY_V4 which is not defined for
      Darwin. ]

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno committed:
docs: Update documentation for BPF IPv6 masquerading

BPF-based IPv6 masquerading is now supported, except for the ip-masq
agent. This commit therefore updates the documentation accordingly.

[ Quentin: Add the additional limitation on IPv6 BPF masquerading with
  host firewall, see previous commit for details. ]

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1722,27 +1722,88 @@ static __always_inline void snat_v6_init_tuple(const struct ipv6hdr *ip6,
+			ipv6_addr_copy(&tuple.daddr, (union v6addr *)&ip6->daddr);
+			ipv6_addr_copy(&tuple.saddr, (union v6addr *)&ip6->saddr);
+
+			l4_off = ETH_HLEN + ipv6_hdrlen(ctx, &ip6->nexthdr);
```
```suggestion
			l4_off = ETH_HLEN + ipv6_hdrlen(ctx, &tuple.nexthdr);
```


to extract the L4 protocol into the CT tuple (for the map lookup), and to avoid that `ipv6_hdrlen()` overwrites the nexthdr field in the IPv6 header.

qmonnet replied:
Right, thanks! Updated

qmonnet committed:
bpf: Change ipv6_addrcmp() into ipv6_addr_equals()

Rename and change function ipv6_addrcmp() into ipv6_addr_equals().
We're not using the comparison result anywhere other than for checking
whether addresses are equal or not (testing whether one is "greater"
than the other does not make much sense for IP addresses, plus the
possible underflow from subtractions in the current implementation would
make it tricky anyway).

Beside improving readability, the motivation for this change is to
address some cases where the verifier rejects programs doing these
subtractions because "math between map_value pointer and register with
unbounded min value is not allowed".

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: Make DEFINE_U*() macros use .data section for declarations

In the bpf package, where we process object files to load programs, we
skip rewriting .bss references for object files without a .data section.
Working on a new feature, this caused some older kernels to complain
with "unrecognized bpf_ld_imm64 insn" (since they don't support
pseudo-register loads), some newer kernels with other errors apparently
related to the same missing references.

A first attempt at fixing the issue consisted in delaying the check for
the presence of the .data section, so that .bss references would be
inlined [0]. This would indeed fix the errors observed in the verifier
tests, but would instead break our BPF unit tests, where .bss is used to
store mock-up value that we want to handle in a specific way.

We need a long-term solution for dealing with the different sections in
a way that works for all set-ups; ideally, we want to be able to use
eBPF global variables everywhere we would need them, but this can't
happen before we drop support for older kernels.

In the meantime, let's adjust macros that can cause declarations to add
variables to the .bss section, in prevision for the next commit. We
explicitly store these variables into .data instead.

[0] https://github.com/cilium/cilium/pull/25837/commits/50f8828510a743e87e18246ea9fe6c96ccaef5e5

Suggested-by: Timo Beckers <timo@isovalent.com>
Suggested-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
test: Test BPF masquerading for IPv6, in conformance-e2e workflow

We know test masquerading in the conformance end-to-end workflow as
well, which should eventually replace Jenkins. Let's update the
workflow's description to use IPv6 in there, as well.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: Rename ENABLE_MASQUERADE as ENABLE_MASQUERADE_IPV4

In preparation for BPF IPv6 masquerading, rename ENABLE_MASQUERADE to
make it IPv4-related. This is because we may enable masquerading for
IPv4 only or IPv6 only in the agent, and in this case, we need to make
a distinction between the two IP versions in the datapath.

All occurrences of the variable are on IPv4-only paths.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: Add ENABLE_MASQUERADE_IPV6 for complexity tests

Add ENABLE_MASQUERADE_IPV6 everywhere we already have
ENABLE_MASQUERADE_IPV4, and ENABLE_IPV6 is set.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
test: net_policies: Disable IPv6 masquerading in addition to IPv4

Now that BPF IPv6 masquerading is supported, align the configuration on
IPv4's for net_policies test, because (as per the comment) "the request
will fail if the reply packet's source address is rewritten when sending
a request directly to the Pod from outside the cluster".

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1764,27 +1764,86 @@ static __always_inline void snat_v6_init_tuple(const struct ipv6hdr *ip6,
 
-		ret = ipv6_hdrlen(ctx, &nexthdr);
-		if (ret > 0) {
-			if (nodeport_uses_dsr(nexthdr))
```
Could you also remove the `nodeport_uses_dsr()` forward declaration at the top of the file? :pray: 

(now I'm looking forward to this PR even more haha)

qmonnet replied:
Yep, done! Thanks

zacharysarah commented on Documentation/network/concepts/masquerading.rst:
```diff
@@ -142,7 +139,7 @@ configure the ``ip-masq-agent`` as above.
 .. note::
 
-    eBPF based masquerading is currently not supported for IPv6 traffic.
+    The eBPF-based ip-masq-agent is currently not supported for IPv6 traffic.
```
I'd like to see "currently" removed, but I'm willing to defer it as out of scope for the PR.

PKizzle replied:
So are we not going to see IPv6 support for ip-masq-agent? Or why do you want to remove "currently"?

qmonnet replied:
My understanding from Sarah's feedback is that having “currently” in the docs is not a good practice - you never really when “currently” refers to (it refers to the date of writing obviously, but when reading the docs, how do you figure this one out?).

Sarah, I'll remove that word for the next iteration. Thanks for the review!

As for the ip-masq-agent with IPv6, it is coming [as a follow-up PR](https://github.com/cilium/cilium/pull/23219), it's _currently_ waiting for this one to be merged.

qmonnet replied:
Now updated in the docs.

zacharysarah reviewed (approved):
LGTM with one fond wish 🥺

tklauser reviewed (approved):
CI bits LGTM sans the test `DO NOT MERGE` commit which I assume will be dropped before merging. I've also briefly skimmed through the other bits in this PR. Nice work!

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1148,13 +1148,13 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, const bool from_host)
 		if (from_host) {
-# if defined(ENABLE_HOST_FIREWALL) && !defined(ENABLE_MASQUERADE)
+# if defined(ENABLE_HOST_FIREWALL) && !defined(ENABLE_MASQUERADE_IPV4)
 			/* If we don't rely on BPF-based masquerading, we need
 			 * to pass the srcid from ipcache to host firewall. See
 			 * comment in ipv4_host_policy_egress() for details.
 			 */
 			ctx_store_meta(ctx, CB_IPCACHE_SRC_LABEL, ipcache_srcid);
-# endif
+# endif /* defined(ENABLE_HOST_FIREWALL) && !defined(ENABLE_MASQUERADE_IPV4) */
```
The asymmetry with the IPv6 path concerns me. But that's not something introduced by your PR :).

qmonnet replied:
What do you mean, having two different paths `ENABLE_MASQUERADE_IPV4` and `...IPV6`? I can't see how to avoid it with the options supported by the Agent, given that you can turn on/off masquerading for one IP version or the other independently :thinking:.

julianwiedmann replied:
It looks like we have special handling for the "no IPv4 BPF masq" case in HostFW, but not yet in the IPv6 path.

I haven't tracked it down further at all, but would speculate that this is because the corresponding IPv6 HostFW path right now **only** has the "no BPF masq" handling. And would need to be extended in similar fashion as IPv4 to also handle BPF masq.

Needs further digging whether that handling for BPF masq in HostFW is an optional optimization, or strictly required.

qmonnet replied:
OK, right, let's disable BPF IPv6 masquerading in conjunction with Host Firewall until we figure out the details here. Thanks a lot for the review and feedback!

julianwiedmann replied:
Thanks! For the record, this is tracked in https://github.com/cilium/cilium/issues/26074.

julianwiedmann commented on bpf/lib/conntrack.h:
```diff
@@ -369,6 +396,31 @@ ipv6_ct_tuple_swap_ports(struct ipv6_ct_tuple *tuple)
+		  enum ct_dir dir __maybe_unused, struct ipv6_ct_tuple *tuple,
+		  bool *has_l4_header __maybe_unused)
+{
+	switch (tuple->nexthdr) {
```
This needs to handle SCTP now as well, correct?

qmonnet replied:
Sounds correct, I added it.

mhofstetter reviewed (approved):
sig-agent part looks good!

julianwiedmann commented on bpf/lib/conntrack.h:
```diff
@@ -369,6 +396,31 @@ ipv6_ct_tuple_swap_ports(struct ipv6_ct_tuple *tuple)
+		break;
+
+	default:
+		/* Can't handle ICMP or extension headers yet. */
```
What does it mean in practice if we don't have ICMPv6 support here? Looks like we wouldn't be able to classify traffic as belonging to an inbound connection, and miss to skip SNAT?

qmonnet replied:
Hmm good question, yes possibly. I added commit https://github.com/cilium/cilium/pull/23165/commits/410aac7ec149a711ae43b954fa837328c79e6853 to refactor a bit based on what Maxim did for IPv4, and to add ICMP support here as well. Can you please take a look?

julianwiedmann replied:
thanks, will stare at it! @gentoo-root FYI

julianwiedmann replied:
looks good, ty!

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1764,27 +1762,86 @@ static __always_inline void snat_v6_init_tuple(const struct ipv6hdr *ip6,
 
 static __always_inline bool snat_v6_needed(struct __ctx_buff *ctx,
-					   const union v6addr *addr)
+					   union v6addr *addr)
```
Did you check whether this is still in sync with `snat_v4_prepare_state()` ? Right now it's a bit difficult to check whether we have all the corresponding cases covered.

I expected them to look slightly different because of EgressGW, but not sure that explains the whole diff :).

qmonnet replied:
I did check. I came up with [a commit](https://github.com/cilium/cilium/commit/243534ea087b84e08c09df3ab10380d31f76df9e) to update `snat_v6_needed()` with respects to `snat_v4_prepare_state()`, but so far I've not included it in the PR.

I noted that the function evolved mostly with the following commits:

  - cdfc30d ("bpf: egressgw: sync logic to determine if destination is outside cluster")
    Moving the `ct_is_reply()` check above the `IPV4_SNAT_EXCLUSION_DST_CIDR` logics makes sense to me, so we can probably apply the non-EGW part of this commit to `snat_v6_needed()`.

  - https://github.com/cilium/cilium/commit/453299601c8c5747deb0b004325db796a2fd31f9 ("bpf: nat: always track egress gateway connections")
    This one is some refactoring which is not necessary at the moment for IPv6, but it seems trivial to apply the few bits touching the relevant function to align with IPv4.

  - https://github.com/cilium/cilium/commit/59ac84f65d2c14fdf8038b29a53b2d1a87f18598 ("bpf: nat: consolidate snat_v4_needed() state into ipv4_nat_target")
    Same.

I probably missed some older commits in the list, but https://github.com/cilium/cilium/commit/243534ea087b84e08c09df3ab10380d31f76df9e brings the two versions (IPv4 and IPv6) pretty close to one another. I'm happy to rework it and add it to the PR.

Let's maybe sync sometimes on Slack/Zoom?

julianwiedmann replied:
Ah nice, thank you! I would need to stare a bit more at your current code to understand if something is genuinely missing, not sure how quickly we could fit in that session :/.

So if you already lined up a sync-up commit for a follow-on, I would probably be fine with merging as-is. And then let's review separately if the sync-up fixes any issues (and needs to go into v1.14), or can wait until v1.15.

qmonnet replied:
Alright, let's keep the current version and make it on par with IPv4 as a follow-up. I'll file an issue to list the different follow-ups to do for this PR.

qmonnet replied:
https://github.com/cilium/cilium/issues/26074

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1764,27 +1762,86 @@ static __always_inline void snat_v6_init_tuple(const struct ipv6hdr *ip6,
+			ipv6_addr_copy(&tuple.daddr, (union v6addr *)&ip6->daddr);
+			ipv6_addr_copy(&tuple.saddr, (union v6addr *)&ip6->saddr);
+
+			l4_off = ETH_HLEN + ipv6_hdrlen(ctx, &tuple.nexthdr);
```
Note that `ipv6_hdrlen()` can return errors.

qmonnet replied:
Good catch. Not addressed yet, I'll need to change the return type of the function to be able to pass an error. Let's see what we do of `snat_v6_needed()` first.

julianwiedmann replied:
FWIW, I think it's fair to only bubble up the error as far as we do in the IPv4 path. And then improve things for both paths in a follow-on.

qmonnet replied:
OK, I'll address by skipping the `ct_is_reply6()` if we get a wrong offset, and leave a comment to improve that in the future.

julianwiedmann commented on bpf/lib/conntrack.h:
```diff
@@ -369,6 +396,31 @@ ipv6_ct_tuple_swap_ports(struct ipv6_ct_tuple *tuple)
+	case IPPROTO_UDP:
+		/* Load sport + dport into tuple. */
+		if (ctx_load_bytes(ctx, off, &tuple->dport, 4) < 0)
+			return DROP_CT_INVALID_HDR;
```
```suggestion
		if (l4_load_ports(ctx, off, &tuple->dport) < 0)
			return DROP_CT_INVALID_HDR;
```

qmonnet replied:
Updated, thanks.

julianwiedmann reviewed (changes requested):
lgtm overall, thanks for working on this Quentin!

One error-handling case that needs to be resolved, and a few clarifying questions below :).

qmonnet committed:
bpf: Add support for ICMP in ct_extract_ports6(), refactor ct_lookup6()

This is the IPv6 port of commit 5ba20bd7e7e5 ("bpf: Use
ct_extract_ports4 for ct_lookup4"). We wouldn't process ICMP in
ct_extract_ports6() so far; this commit moves the more complete
switch/case block from ct_lookup6() to ct_extract_ports6() instead, and
makes the former function call the latter. This way, we get less
duplicated code and ICMP support in ct_extract_ports6().

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
daemon: Forbid IPv6 BPF masquerading with the host firewall

The host firewall had some code paths related to BPF masquerading for
IPv4; now that we added IPv6 support for BPF masquerading, we need to
take a deep look at the implications for this part of the datapath.
Until we have done so, reject IPv6 BPF masquerading in the Agent.

Also disable IPv6 BPF masquerading in some tests involving on the host
firewall.

Suggested-by: Julian Wiedmann <jwi@isovalent.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

julianwiedmann reviewed (approved):
Checks out, and the remaining follow-on items are ok. Thank you!
